File: Reverse a Linked List in groups of given size. - GFG/reverse-a-linked-list-in-groups-of-given-size..cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


struct node
{
    int data;
    struct node* next;
    
    node(int x){
        data = x;
        next = NULL;
    }
    
};

/* Function to print linked list */
void printList(struct node *node)
{
    while (node != NULL)
    {
        printf("%d ", node->data);
        node = node->next;
    }
    printf("\n");
}


 // } Driver Code Ends
/*
  Reverse a linked list
  The input list will have at least one element  
  Return the node which points to the head of the new LinkedList
  Node is defined as 
    struct node
    {
        int data;
        struct node* next;
    
        node(int x){
            data = x;
            next = NULL;
        }
    
    }*head;
*/

class Solution
{
    public:
    node*tt=NULL;
    node*th=NULL;
    void helper(node*ele)
    {
        if(th==NULL)
        {
            th=ele;
            tt=ele;
        }
        else
        {
            ele->next=th;
            th=ele;
        }
    }
    struct node *reverse (struct node *head, int k)
    { 
        node*newHead=new node(-1);
        node*temp=newHead;
        while(head)
        {
            for(int i=0;i<k &&head;i++)
            {
                node*temp=head->next;
                head->next=NULL;
                helper(head);
                head=temp;
            }
            temp->next=th;
            temp=tt;
            tt=NULL;
            th=NULL;
        }return newHead->next;
    }
};


// { Driver Code Starts.

/* Drier program to test above function*/
int main(void)
{
    int t;
    cin>>t;
     
    while(t--)
    {
        struct node* head = NULL;
        struct node* temp = NULL;
        int n;
        cin >> n;
         
        for(int i=0 ; i<n ; i++)
        {
            int value;
            cin >> value;
            if(i == 0)
            {
                head = new node(value);
                temp = head;
            }
            else
            {
                temp->next = new node(value);
                temp = temp->next;
            }
        }
        
        int k;
        cin>>k;
        
        Solution ob;
        head = ob.reverse(head, k);
        printList(head);
    }
     
    return(0);
}

  // } Driver Code Ends

File: Reverse a Linked List in groups of given size. - GFG/reverse-a-linked-list-in-groups-of-given-size..java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;

class Node
{
    int data;
    Node next;
    Node(int key)
    {
        data = key;
        next = null;
    }
}

class ReverseInSize
{
    static Node head;
    
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            int n = sc.nextInt();
            int a1 = sc.nextInt();
            Node head = new Node(a1);
            Node tail = head;
            for(int i = 1; i < n; i++)
            {
                int a = sc.nextInt();
                // addToTheLast(new Node(a));
                tail.next = new Node(a);
                tail =tail.next;
            }
            
            int k = sc.nextInt();
            Solution ob = new Solution();
            Node res = ob.reverse(head, k);
            printList(res);
            System.out.println();
        }
    }
    
    public static void printList(Node node)
    {
        while(node != null)
        {
            System.out.print(node.data + " ");
            node = node.next;
        }
    }
    
}


// } Driver Code Ends


/*node class of the linked list

class Node
{
    int data;
    Node next;
    Node(int key)
    {
        data = key;
        next = null;
    }
}

*/

class Solution
{
    static Node th=null;
    static Node tt=null;
    public static void addOne(Node ele)
    {
        if(th==null)
        {
            th=tt=ele;
        }
        else{
            ele.next=th;
            th=ele;
        }
    }
    
    public static Node reverse(Node node, int k)
    {
        Node newHead=new Node(-1);
        Node trail=newHead;
        Node head=node;
        while(head!=null)
        {
            for(int i=0;i<k && head!=null;i++)
            {
                Node temp=head.next;
                head.next=null;
                addOne(head);
                head=temp;
            }
            trail.next=th;
            trail=tt;
            tt=th=null;
        }
        return newHead.next;
    }
}

